File: opencode-config/skills/evo-discover/references/inline_instrumentation.py
"""Inline instrumentation template for Python benchmarks.

Use this when the user declines the `evo-agent` SDK. Paste the helper into
the benchmark script and call `log_task()` + `write_result()` in place of
the SDK's `Run` context manager. Zero new dependencies.

Contract (same as the SDK):
- Read EVO_TRACES_DIR and EVO_EXPERIMENT_ID from the environment.
- Write task_<id>.json files into EVO_TRACES_DIR as each task finishes.
- Print a single JSON object with a "score" field to stdout at the end.
- All other output goes to stderr.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

_TRACES_DIR = Path(os.environ["EVO_TRACES_DIR"]) if os.environ.get("EVO_TRACES_DIR") else None
_EXPERIMENT_ID = os.environ.get("EVO_EXPERIMENT_ID", "unknown")
_SCORES: dict[str, float] = {}
_STARTED_AT = datetime.now(timezone.utc).isoformat(timespec="seconds")

if _TRACES_DIR:
    _TRACES_DIR.mkdir(parents=True, exist_ok=True)


def log_task(
    task_id: str,
    score: float,
    *,
    summary: str | None = None,
    failure_reason: str | None = None,
    log: list[Any] | None = None,
    **extra: Any,
) -> None:
    """Record the result for one task. Writes task_<id>.json immediately."""
    task_id = str(task_id)
    _SCORES[task_id] = score
    if _TRACES_DIR is None:
        return
    trace: dict[str, Any] = {
        "experiment_id": _EXPERIMENT_ID,
        "task_id": task_id,
        "status": "passed" if score >= 0.5 else "failed",
        "score": score,
        "ended_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }
    if summary is not None:
        trace["summary"] = summary
    if failure_reason is not None:
        trace["failure_reason"] = failure_reason
    if log is not None:
        trace["log"] = log
    trace.update(extra)
    (_TRACES_DIR / f"task_{task_id}.json").write_text(
        json.dumps(trace, indent=2), encoding="utf-8"
    )


def write_result(score: float | None = None) -> float:
    """Emit the final score JSON to stdout and return the score.

    The return value lets callers implement --min-score gate logic without
    recomputing the aggregate.
    """
    if score is None:
        score = sum(_SCORES.values()) / len(_SCORES) if _SCORES else 0.0
    score = round(score, 4)
    result = {
        "score": score,
        "tasks": dict(_SCORES),
        "started_at": _STARTED_AT,
        "ended_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
    }
    print(json.dumps(result, indent=2))
    return score

File: opencode-config/skills/evo-discover/references/inline_instrumentation.js
/**
 * Inline instrumentation template for Node benchmarks.
 *
 * Use this when the user declines the `evo-agent` SDK. Import the helpers
 * into the benchmark script and call logTask() + writeResult() in place of
 * the SDK's Run class. Zero new dependencies.
 *
 * Contract (same as the SDK):
 * - Read EVO_TRACES_DIR and EVO_EXPERIMENT_ID from process.env.
 * - Write task_<id>.json files into EVO_TRACES_DIR as each task finishes.
 * - Print a single JSON object with a "score" field to stdout at the end.
 * - All other output goes to stderr.
 */

import { writeFileSync, mkdirSync } from "node:fs";
import { join } from "node:path";

const TRACES_DIR = process.env.EVO_TRACES_DIR || null;
const EXPERIMENT_ID = process.env.EVO_EXPERIMENT_ID || "unknown";
const SCORES = {};
const STARTED_AT = new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");

if (TRACES_DIR) mkdirSync(TRACES_DIR, { recursive: true });

export function logTask(taskId, score, { summary, failureReason, log, ...extra } = {}) {
  taskId = String(taskId);
  SCORES[taskId] = score;
  if (!TRACES_DIR) return;
  const trace = {
    experiment_id: EXPERIMENT_ID,
    task_id: taskId,
    status: score >= 0.5 ? "passed" : "failed",
    score,
    ended_at: new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00"),
  };
  if (summary !== undefined) trace.summary = summary;
  if (failureReason !== undefined) trace.failure_reason = failureReason;
  if (log !== undefined) trace.log = log;
  Object.assign(trace, extra);
  writeFileSync(join(TRACES_DIR, `task_${taskId}.json`), JSON.stringify(trace, null, 2), "utf-8");
}

export function writeResult(score) {
  const ids = Object.keys(SCORES);
  if (score === undefined) {
    score = ids.length === 0 ? 0.0 : ids.reduce((a, id) => a + SCORES[id], 0) / ids.length;
  }
  score = Math.round(score * 10000) / 10000;
  const result = {
    score,
    tasks: { ...SCORES },
    started_at: STARTED_AT,
    ended_at: new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00"),
  };
  process.stdout.write(JSON.stringify(result, null, 2) + "\n");
  return score;
}
